pkgs/by-name/sh/share-jvm/hook.sh

```diff
@@ -0,0 +1,12 @@
+# shellcheck shell=bash
+
+postInstallHooks+=(shareJvm)
+
+shareJvm() {
+    if [ "${dontShareJvm-}" = 1 ]; then return; fi
+
+    if [ ! -d "${!outputBin:?}" ]; then return; fi
+
+    mkdir -p "${!outputBin:?}/share/jvm"
+    ln -fs "${!outputBin:?}" "${!outputBin:?}/share/jvm/${name}"
```

Comment:
I would rather opt for `lib`, which is usually meant for platform specific things, and is already more of a convention in Gradle:

> [/usr/lib{,64}/jvm, {/opt,/usr/{,local/}}java](https://github.com/gradle/gradle/blob/766a6288794e4f5b78ad44e9ba8b0a2204d3c439/platforms/jvm/jvm-services/src/main/java/org/gradle/jvm/toolchain/internal/LinuxInstallationSupplier.java#L36)

```suggestion
    mkdir -p "${!outputBin:?}/lib/jvm"
    ln -fs "${!outputBin:?}" "${!outputBin:?}/share/lib/${name}"
```


The `share` is described in [Wiki about FHS](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard) as:

```
Architecture-independent (shared) data.
```

And, JVM is architecture dependent.

---

Though, IMO, `lib` isn't totally correct either. As JVM is collection of binaries and libraries. But, it's an already adopted pattern.

Reply:
I have skimmed [FHS 2.3](https://refspecs.linuxfoundation.org/FHS_2.3/fhs-2.3.pdf) before choosing `$out/share/jvm`. It reads like from another planet. I think that `/lib` is wrong, `/share` is also wrong, and the FHS won't really help you because it is just not written to spec what we are doing here.

`/share` is (ab)used by many others: https://github.com/NixOS/nixpkgs/blob/e7cc61784ddf51c81487637b3031a6dd2d6673a2/nixos/modules/config/system-path.nix#L154-L165

Reply:
`lib` canonically doesn't have to consist of *only* binaries. `pkgconfig` files go in there, for example. The rule of thumb is architecture-dependent or not, and I think all the things I recognize in that system-path.nix excerpt are architecture-independent. A JVM isn't.

Reply:
Hmm, the biggest reason for me to avoid `/lib` was the corresponding comment suggesting that it could be removed:

https://github.com/NixOS/nixpkgs/blob/e7cc61784ddf51c81487637b3031a6dd2d6673a2/nixos/modules/config/system-path.nix#L152

However this FIXME has been sitting there for about ~10 years, see https://github.com/NixOS/nixpkgs/commit/e636e0a5329f546ebcc603d56bb99a0d016370f8#diff-5c7b357730b03feb709650ecc8c17df225af3b31ad01b5142de81b33c4f72639R106

So maybe `$out/lib` would also be okay. If this somehow blocks merging, I'm happy to go for `$out/lib`. But I'd be very curious to understand why `/lib` should be removed.

Reply:
Alright, I am changing to `$out/lib/jvm`. @rhendric can you enlighten me regarding the FIXME?

Reply:
> @rhendric can you enlighten me regarding the FIXME?

:sweat_smile: I have no idea what, if anything, is wrong with linking `/lib`; looks like an @edolstra -ism.

A theory is that, since the [file referenced](https://github.com/NixOS/nixpkgs/blob/40c21be908c26dc2b1bc749cde92f46cac92a3a2/nixos/modules/config/debug-info.nix) refers to wanting to link a specific subpath under `/lib`, if the `/lib` entry were to be removed for some reason it would still be correct to add `/lib/jvm` back in.